### PR TITLE
fix window functions for Distributed tables

### LINKED ISSUE #22121: Window functions: wrong result with Distributed table
**Describe the bug**
Window functions return wrong result when used with Distributed table that contains 3 shards and uses
random sharding key.

**How to reproduce**
```
                 1ms                  [clickhouse1] bash# echo -e "
                 3ms                  [clickhouse1] >             CREATE TABLE empsalary_source ON CLUSTER sharded_cluster (
                 5ms                  [clickhouse1] >                 depname LowCardinality(String),
                 6ms                  [clickhouse1] >                 empno  UInt64,
                 8ms                  [clickhouse1] >                 salary Int32,
                 9ms                  [clickhouse1] >                 enroll_date Date
                11ms                  [clickhouse1] >                 )
                13ms                  [clickhouse1] >             ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/empsalary_source', '{replica}') ORDER BY enroll_date
                16ms                  [clickhouse1] >             " | clickhouse client -n --allow_experimental_window_functions "1"
               181ms                  [clickhouse1] clickhouse3	9000	0		2	0
               181ms                  [clickhouse1] clickhouse1	9000	0		1	0
               237ms                  [clickhouse1] clickhouse2	9000	0		0	0
               244ms                  [clickhouse1] bash# echo $?
               244ms                  [clickhouse1] 0
               245ms                  [clickhouse1] bash#
                 2ms                  [clickhouse1] bash# echo -e "CREATE TABLE empsalary AS empsalary_source ENGINE = Distributed(sharded_cluster, default, empsalary_source, rand())" | clickhouse client -n --allow_experimental_window_functions "1"
                36ms                  [clickhouse1] bash# echo $?
                37ms                  [clickhouse1] 0
                37ms                  [clickhouse1] bash#
                 3ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('develop', 10, 5200, '2007-08-01')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                49ms                [clickhouse1] bash# echo $?
                49ms                [clickhouse1] 0
                50ms                [clickhouse1] bash#
                 2ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('sales', 1, 5000, '2006-10-01')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                41ms                [clickhouse1] bash# echo $?
                42ms                [clickhouse1] 0
                42ms                [clickhouse1] bash#
                 3ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('personnel', 5, 3500, '2007-12-10')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                51ms                [clickhouse1] bash# echo $?
                52ms                [clickhouse1] 0
                52ms                [clickhouse1] bash#
                 3ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('sales', 4, 4800, '2007-08-08')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                42ms                [clickhouse1] bash# echo $?
                42ms                [clickhouse1] 0
                42ms                [clickhouse1] bash#
                 3ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('personnel', 2, 3900, '2006-12-23')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                42ms                [clickhouse1] bash# echo $?
                42ms                [clickhouse1] 0
                42ms                [clickhouse1] bash#
                 2ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('develop', 7, 4200, '2008-01-01')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                39ms                [clickhouse1] bash# echo $?
                39ms                [clickhouse1] 0
                39ms                [clickhouse1] bash#
                 3ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('develop', 9, 4500, '2008-01-01')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                38ms                [clickhouse1] bash# echo $?
                38ms                [clickhouse1] 0
                38ms                [clickhouse1] bash#
                 3ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('sales', 3, 4800, '2007-08-01')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                39ms                [clickhouse1] bash# echo $?
                39ms                [clickhouse1] 0
                40ms                [clickhouse1] bash#
                 2ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('develop', 8, 6000, '2006-10-01')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                36ms                [clickhouse1] bash# echo $?
                36ms                [clickhouse1] 0
                36ms                [clickhouse1] bash#
                 3ms                [clickhouse1] bash# echo -e "INSERT INTO empsalary VALUES ('develop', 11, 5200, '2007-08-15')" | clickhouse client -n --insert_distributed_sync "1" --allow_experimental_window_functions "1"
                43ms                [clickhouse1] bash# echo $?
                43ms                [clickhouse1] 0
                44ms                [clickhouse1] bash#
                 2ms                [clickhouse1] bash# echo -e "SELECT * FROM empsalary" | clickhouse client -n --allow_experimental_window_functions "1"
                22ms                [clickhouse1] sales	3	4800	2007-08-01
                23ms                [clickhouse1] personnel	5	3500	2007-12-10
                38ms                [clickhouse1] sales	4	4800	2007-08-08
                38ms                [clickhouse1] develop	8	6000	2006-10-01
                39ms                [clickhouse1] develop	7	4200	2008-01-01
                39ms                [clickhouse1] develop	10	5200	2007-08-01
                39ms                [clickhouse1] develop	9	4500	2008-01-01
                39ms                [clickhouse1] personnel	2	3900	2006-12-23
                39ms                [clickhouse1] develop	11	5200	2007-08-15
                40ms                [clickhouse1] sales	1	5000	2006-10-01
                47ms                [clickhouse1] bash# echo $?
                48ms                [clickhouse1] 0
                48ms                [clickhouse1] bash#
                 2ms                [clickhouse1] bash# echo -e "SELECT depname, empno, salary, sum(salary) OVER (PARTITION BY depname) AS sum FROM empsalary ORDER BY depname, salary FORMAT TabSeparatedWithNames" | clickhouse client -n --allow_experimental_window_functions "1"
                26ms                [clickhouse1] depname	empno	salary	sum
                31ms                [clickhouse1] develop	7	4200	13900
                31ms                [clickhouse1] develop	9	4500	13900
                31ms                [clickhouse1] develop	10	5200	11200
                32ms                [clickhouse1] develop	11	5200	13900
                32ms                [clickhouse1] develop	8	6000	11200
                32ms                [clickhouse1] personnel	5	3500	3500
                32ms                [clickhouse1] personnel	2	3900	3900
                32ms                [clickhouse1] sales	4	4800	9600
                32ms                [clickhouse1] sales	3	4800	9600
                32ms                [clickhouse1] sales	1	5000	5000
                39ms                [clickhouse1] bash# echo $?
                39ms                [clickhouse1] 0
                39ms                [clickhouse1] bash#
                 2ms                [clickhouse1] bash# echo -e "DROP TABLE IF EXISTS empsalary" | clickhouse client -n --allow_experimental_window_functions "1"
                27ms                [clickhouse1] bash# echo $?
                27ms                [clickhouse1] 0
                27ms                [clickhouse1] bash#
                 3ms                [clickhouse1] bash# echo -e "DROP TABLE IF EXISTS empsalary_source ON CLUSTER sharded_cluster" | clickhouse client -n --allow_experimental_window_functions "1"
               143ms                [clickhouse1] clickhouse2	9000	0		2	0
               143ms                [clickhouse1] clickhouse3	9000	0		1	0
               143ms                [clickhouse1] clickhouse1	9000	0		0	0
               150ms                [clickhouse1] bash# echo $?
               150ms                [clickhouse1] 0
               151ms                [clickhouse1] bash#
               171ms                [clickhouse1] 

```

**Expected behavior**
Expected output

```
      depname  | empno | salary |  sum  
    -----------+-------+--------+-------
     develop   |     7 |   4200 | 25100
     develop   |     9 |   4500 | 25100
     develop   |    10 |   5200 | 25100
     develop   |    11 |   5200 | 25100
     develop   |     8 |   6000 | 25100
     personnel |     5 |   3500 |  7400
     personnel |     2 |   3900 |  7400
     sales     |     4 |   4800 | 14600
     sales     |     3 |   4800 | 14600
     sales     |     1 |   5000 | 14600
```

### PR DESCRIPTION
**Pull request by akuzm**
They are probably still broken but somewhat less...

This PR moves the calculation of window functions to the initiator. They were calculated on shards before, which is generally incorrect.

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Fixes https://github.com/ClickHouse/ClickHouse/issues/22121

### CODE REVIEW

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -627,6 +627,10 @@ Block InterpreterSelectQuery::getSampleBlockImpl()
         if (!analysis_result.need_aggregate)
         {
             // What's the difference with selected_columns?
```
Comment (KochetovNicolai) [inline src/Interpreters/InterpreterSelectQuery.cpp:629]: It is the header we want from remote server, to execute query up to WithMergeableState
When there is an ORDER BY, it is executed on remote server firstly, then we execute merge sort on initiator.
To execute ORDER BY, we need to calc ORDER BY keys.

**Review by KochetovNicolai — APPROVED** [review]
LGTM

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 1 above)
Comment (akuzm) [reply src/Interpreters/InterpreterSelectQuery.cpp:629]: Added a comment.

**Comment by akuzm** [comment]
Integration tests are broken in master, and Performance and PVS failed because of the improvements I am continuously inflicting on our testing infractructure.

**Pull request merged**